File: infra/microsoft-copilot/semantic-kernel/src/ForgeSquad.Core/Pipeline/ApprovalGate.cs
using System.Net.Http.Json;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;

namespace ForgeSquad.Core.Pipeline;

/// <summary>
/// Manages human-in-the-loop approval gates for pipeline checkpoints.
/// Supports two modes:
/// 1. Microsoft Teams Approvals (production) — sends Adaptive Card, waits for response
/// 2. Console/API mode (development) — waits for input via console or REST API
/// </summary>
public sealed class ApprovalGate
{
    private readonly HttpClient _httpClient;
    private readonly ApprovalGateOptions _options;
    private readonly ILogger<ApprovalGate> _logger;

    public ApprovalGate(
        HttpClient httpClient,
        IOptions<ApprovalGateOptions> options,
        ILogger<ApprovalGate> logger)
    {
        _httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
        _options = options?.Value ?? throw new ArgumentNullException(nameof(options));
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
    }

    /// <summary>
    /// Requests approval from designated approvers and waits for a decision.
    /// </summary>
    public async Task<ApprovalResult> RequestApprovalAsync(
        ApprovalRequest request,
        CancellationToken cancellationToken = default)
    {
        _logger.LogInformation(
            "Requesting approval for checkpoint {CP} (phase: {Phase}, run: {Run})",
            request.CheckpointNumber, request.PhaseName, request.RunId);

        // Validate artifact integrity
        var computedHash = ComputeSha256(request.ArtifactSummary);
        if (!string.Equals(computedHash, request.ArtifactHash, StringComparison.OrdinalIgnoreCase))
        {
            _logger.LogError("Artifact hash mismatch at checkpoint {CP}. Expected: {Expected}, Got: {Got}",
                request.CheckpointNumber, request.ArtifactHash, computedHash);

            return new ApprovalResult
            {
                Decision = ApprovalDecision.Rejected,
                DecidedBy = "System",
                Comments = "SECURITY: Artifact hash mismatch detected. Artifacts may have been tampered with.",
                DecidedAt = DateTimeOffset.UtcNow,
                ArtifactHash = computedHash
            };
        }

        return _options.Mode switch
        {
            ApprovalMode.Teams => await RequestTeamsApprovalAsync(request, cancellationToken),
            ApprovalMode.Console => await RequestConsoleApprovalAsync(request, cancellationToken),
            ApprovalMode.Api => await RequestApiApprovalAsync(request, cancellationToken),
            ApprovalMode.AutoApprove => CreateAutoApproval(request),
            _ => throw new InvalidOperationException($"Unknown approval mode: {_options.Mode}")
        };
    }

    /// <summary>
    /// Sends an approval request via Microsoft Teams Adaptive Card and waits for response.
    /// Uses the Teams Approvals API via Power Automate webhook.
    /// </summary>
    private async Task<ApprovalResult> RequestTeamsApprovalAsync(
        ApprovalRequest request,
        CancellationToken cancellationToken)
    {
        var adaptiveCard = BuildAdaptiveCard(request);

        // Create approval via Teams webhook (Power Automate)
        var payload = new
        {
            type = "AdaptiveCard",
            title = $"ForgeSquad Checkpoint {request.CheckpointNumber}: {request.PhaseName}",
            assignedTo = string.Join(";", request.Approvers),
            details = request.ArtifactSummary,
            adaptiveCard,
            runId = request.RunId,
            checkpointNumber = request.CheckpointNumber,
            artifactHash = request.ArtifactHash
        };

        _logger.LogInformation("Sending Teams approval request for checkpoint {CP}", request.CheckpointNumber);

        var response = await _httpClient.PostAsJsonAsync(
            _options.TeamsWebhookUrl, payload, cancellationToken);
        response.EnsureSuccessStatusCode();

        // Poll for approval response
        var timeout = TimeSpan.FromMinutes(request.TimeoutMinutes);
        var pollInterval = TimeSpan.FromSeconds(30);
        var deadline = DateTimeOffset.UtcNow + timeout;

        while (DateTimeOffset.UtcNow < deadline)
        {
            cancellationToken.ThrowIfCancellationRequested();

            var statusResponse = await _httpClient.GetAsync(
                $"{_options.ApprovalStatusUrl}?runId={request.RunId}&checkpoint={request.CheckpointNumber}",
                cancellationToken);

            if (statusResponse.IsSuccessStatusCode)
            {
                var status = await statusResponse.Content.ReadFromJsonAsync<ApprovalStatusResponse>(
                    cancellationToken: cancellationToken);

                if (status?.IsCompleted == true)
                {
                    _logger.LogInformation("Checkpoint {CP} decided: {Decision} by {Approver}",
                        request.CheckpointNumber, status.Decision, status.DecidedBy);

                    return new ApprovalResult
                    {
                        Decision = ParseDecision(status.Decision),
                        DecidedBy = status.DecidedBy ?? "Unknown",
                        Comments = status.Comments ?? string.Empty,
                        DecidedAt = DateTimeOffset.UtcNow,
                        ArtifactHash = request.ArtifactHash
                    };
                }
            }

            await Task.Delay(pollInterval, cancellationToken);
        }

        _logger.LogWarning("Checkpoint {CP} timed out after {Minutes} minutes",
            request.CheckpointNumber, request.TimeoutMinutes);

        return new ApprovalResult
        {
            Decision = ApprovalDecision.Rejected,
            DecidedBy = "System (Timeout)",
            Comments = $"Approval timed out after {request.TimeoutMinutes} minutes",
            DecidedAt = DateTimeOffset.UtcNow,
            ArtifactHash = request.ArtifactHash
        };
    }

    /// <summary>
    /// Requests approval via console input (for development/testing).
    /// </summary>
    private async Task<ApprovalResult> RequestConsoleApprovalAsync(
        ApprovalRequest request,
        CancellationToken cancellationToken)
    {
        Console.WriteLine();
        Console.WriteLine("=".PadRight(60, '='));
        Console.WriteLine($"  CHECKPOINT {request.CheckpointNumber}: {request.PhaseName}");
        Console.WriteLine("=".PadRight(60, '='));
        Console.WriteLine();
        Console.WriteLine("Artifact Summary:");
        Console.WriteLine(request.ArtifactSummary[..Math.Min(1000, request.ArtifactSummary.Length)]);
        Console.WriteLine();
        Console.WriteLine($"Artifact Hash (SHA-256): {request.ArtifactHash}");
        Console.WriteLine();
        Console.WriteLine("Decision: [A]pprove / [R]eject / [M]odify");
        Console.Write("> ");

        var input = Console.ReadLine()?.Trim().ToUpperInvariant();
        var decision = input switch
        {
            "A" or "APPROVE" => ApprovalDecision.Approved,
            "R" or "REJECT" => ApprovalDecision.Rejected,
            "M" or "MODIFY" => ApprovalDecision.Modified,
            _ => ApprovalDecision.Approved // Default to approved in dev mode
        };

        Console.Write("Comments (optional): ");
        var comments = Console.ReadLine() ?? string.Empty;

        return new ApprovalResult
        {
            Decision = decision,
            DecidedBy = "Console User",
            Comments = comments,
            DecidedAt = DateTimeOffset.UtcNow,
            ArtifactHash = request.ArtifactHash
        };
    }

    /// <summary>
    /// Requests approval via REST API (for programmatic integration).
    /// Posts the request and polls for response.
    /// </summary>
    private async Task<ApprovalResult> RequestApiApprovalAsync(
        ApprovalRequest request,
        CancellationToken cancellationToken)
    {
        // Post approval request to external API
        var response = await _httpClient.PostAsJsonAsync(
            $"{_options.ApiBaseUrl}/approvals",
            new
            {
                runId = request.RunId,
                checkpointNumber = request.CheckpointNumber,
                phaseName = request.PhaseName,
                artifactSummary = request.ArtifactSummary,
                artifactHash = request.ArtifactHash,
                approvers = request.Approvers,
                timeoutMinutes = request.TimeoutMinutes
            },
            cancellationToken);

        response.EnsureSuccessStatusCode();

        // Poll for decision
        var timeout = TimeSpan.FromMinutes(request.TimeoutMinutes);
        var deadline = DateTimeOffset.UtcNow + timeout;

        while (DateTimeOffset.UtcNow < deadline)
        {
            cancellationToken.ThrowIfCancellationRequested();

            var statusResponse = await _httpClient.GetFromJsonAsync<ApprovalStatusResponse>(
                $"{_options.ApiBaseUrl}/approvals/{request.RunId}/{request.CheckpointNumber}",
                cancellationToken);

            if (statusResponse?.IsCompleted == true)
            {
                return new ApprovalResult
                {
                    Decision = ParseDecision(statusResponse.Decision),
                    DecidedBy = statusResponse.DecidedBy ?? "API User",
                    Comments = statusResponse.Comments ?? string.Empty,
                    DecidedAt = DateTimeOffset.UtcNow,
                    ArtifactHash = request.ArtifactHash
                };
            }

            await Task.Delay(TimeSpan.FromSeconds(10), cancellationToken);
        }

        return new ApprovalResult
        {
            Decision = ApprovalDecision.Rejected,
            DecidedBy = "System (Timeout)",
            Comments = "API approval timed out",
            DecidedAt = DateTimeOffset.UtcNow,
            ArtifactHash = request.ArtifactHash
        };
    }

    /// <summary>
    /// Auto-approves (for testing/CI pipelines).
    /// </summary>
    private static ApprovalResult CreateAutoApproval(ApprovalRequest request)
    {
        return new ApprovalResult
        {
            Decision = ApprovalDecision.Approved,
            DecidedBy = "AutoApprove (Test Mode)",
            Comments = "Automatically approved in test mode",
            DecidedAt = DateTimeOffset.UtcNow,
            ArtifactHash = request.ArtifactHash
        };
    }

    /// <summary>
    /// Builds a Teams Adaptive Card for the approval request.
    /// </summary>
    private static object BuildAdaptiveCard(ApprovalRequest request)
    {
        return new
        {
            type = "AdaptiveCard",
            version = "1.5",
            body = new object[]
            {
                new
                {
                    type = "TextBlock",
                    text = $"ForgeSquad Checkpoint {request.CheckpointNumber}",
                    weight = "Bolder",
                    size = "Large"
                },
                new
                {
                    type = "TextBlock",
                    text = $"Phase: {request.PhaseName}",
                    weight = "Bolder"
                },
                new
                {
                    type = "TextBlock",
                    text = $"Run ID: {request.RunId}",
                    isSubtle = true
                },
                new
                {
                    type = "TextBlock",
                    text = "Artifact Summary:",
                    weight = "Bolder",
                    separator = true
                },
                new
                {
                    type = "TextBlock",
                    text = request.ArtifactSummary.Length > 500
                        ? request.ArtifactSummary[..500] + "..."
                        : request.ArtifactSummary,
                    wrap = true
                },
                new
                {
                    type = "FactSet",
                    facts = new[]
                    {
                        new { title = "Artifact Hash", value = request.ArtifactHash[..16] + "..." },
                        new { title = "Timeout", value = $"{request.TimeoutMinutes} minutes" }
                    }
                }
            },
            actions = new object[]
            {
                new
                {
                    type = "Action.Submit",
                    title = "Approve",
                    style = "positive",
                    data = new { decision = "Approved", checkpointNumber = request.CheckpointNumber }
                },
                new
                {
                    type = "Action.Submit",
                    title = "Reject",
                    style = "destructive",
                    data = new { decision = "Rejected", checkpointNumber = request.CheckpointNumber }
                },
                new
                {
                    type = "Action.Submit",
                    title = "Request Changes",
                    data = new { decision = "Modified", checkpointNumber = request.CheckpointNumber }
                }
            }
        };
    }

    private static string ComputeSha256(string content)
    {
        var bytes = SHA256.HashData(Encoding.UTF8.GetBytes(content));
        return Convert.ToHexString(bytes).ToLowerInvariant();
    }

    private static ApprovalDecision ParseDecision(string? decision)
    {
        return decision?.ToLowerInvariant() switch
        {
            "approved" or "approve" => ApprovalDecision.Approved,
            "rejected" or "reject" => ApprovalDecision.Rejected,
            "modified" or "modify" or "request changes" => ApprovalDecision.Modified,
            _ => ApprovalDecision.Rejected
        };
    }
}

// --- Data models ---

public sealed class ApprovalRequest
{
    public required string RunId { get; init; }
    public required int CheckpointNumber { get; init; }
    public required string PhaseName { get; init; }
    public required string ArtifactSummary { get; init; }
    public required string ArtifactHash { get; init; }
    public required List<string> Approvers { get; init; }
    public int TimeoutMinutes { get; init; } = 1440; // 24 hours default
}

public sealed class ApprovalResult
{
    public ApprovalDecision Decision { get; init; }
    public string DecidedBy { get; init; } = string.Empty;
    public string Comments { get; init; } = string.Empty;
    public DateTimeOffset DecidedAt { get; init; }
    public string ArtifactHash { get; init; } = string.Empty;
}

public sealed class ApprovalStatusResponse
{
    [JsonPropertyName("isCompleted")]
    public bool IsCompleted { get; set; }

    [JsonPropertyName("decision")]
    public string? Decision { get; set; }

    [JsonPropertyName("decidedBy")]
    public string? DecidedBy { get; set; }

    [JsonPropertyName("comments")]
    public string? Comments { get; set; }
}

public sealed class ApprovalGateOptions
{
    public ApprovalMode Mode { get; set; } = ApprovalMode.Console;
    public string? TeamsWebhookUrl { get; set; }
    public string? ApprovalStatusUrl { get; set; }
    public string? ApiBaseUrl { get; set; }
}

public enum ApprovalMode
{
    Teams,
    Console,
    Api,
    AutoApprove
}

public enum ApprovalDecision
{
    Approved,
    Rejected,
    Modified
}

File: infra/microsoft-copilot/autogen/agents/approval_gate.py
"""
ForgeSquad — Human-in-the-Loop Approval Gate for AutoGen

Implements checkpoint management with support for:
1. Console mode — interactive approval via terminal
2. Teams mode — sends notification via webhook, polls for response
3. Auto-approve mode — for testing and CI pipelines
"""

import hashlib
import logging
import os
from dataclasses import dataclass, field
from datetime import datetime, timezone

import httpx

logger = logging.getLogger(__name__)


@dataclass
class ApprovalRequest:
    """Request for human approval at a pipeline checkpoint."""
    run_id: str
    checkpoint_number: int
    phase_name: str
    artifact_summary: str
    artifact_hash: str
    approvers: list[str] = field(default_factory=lambda: ["admin@company.com"])
    timeout_minutes: int = 1440  # 24 hours


@dataclass
class ApprovalResult:
    """Result of an approval decision."""
    decision: str  # "approved", "rejected", "modified"
    decided_by: str
    comments: str
    decided_at: str
    artifact_hash: str


class ApprovalGate:
    """
    Manages human-in-the-loop approval gates for ForgeSquad pipeline checkpoints.

    Supports three modes:
    - console: Interactive terminal input (development)
    - teams: Microsoft Teams webhook notifications (production)
    - auto_approve: Automatic approval (testing/CI)
    """

    def __init__(
        self,
        mode: str | None = None,
        auto_approve: bool = False,
        teams_webhook_url: str | None = None,
        approval_api_url: str | None = None,
    ):
        if auto_approve:
            self.mode = "auto_approve"
        else:
            self.mode = mode or os.getenv("APPROVAL_MODE", "console")

        self.teams_webhook_url = (
            teams_webhook_url or os.getenv("TEAMS_WEBHOOK_URL", "")
        )
        self.approval_api_url = (
            approval_api_url or os.getenv("APPROVAL_API_URL", "")
        )

    async def request_approval(self, request: ApprovalRequest) -> ApprovalResult:
        """
        Requests human approval for a checkpoint.

        First validates artifact integrity via SHA-256 hash,
        then delegates to the appropriate approval method.
        """
        # Validate artifact integrity
        computed_hash = hashlib.sha256(
            request.artifact_summary.encode()
        ).hexdigest()

        if computed_hash != request.artifact_hash:
            logger.error(
                "Artifact hash mismatch at checkpoint %d! "
                "Expected: %s, Got: %s",
                request.checkpoint_number,
                request.artifact_hash,
                computed_hash,
            )
            return ApprovalResult(
                decision="rejected",
                decided_by="System (Integrity Check)",
                comments="SECURITY: Artifact hash mismatch. Possible tampering detected.",
                decided_at=datetime.now(timezone.utc).isoformat(),
                artifact_hash=computed_hash,
            )

        if self.mode == "auto_approve":
            return self._auto_approve(request)
        elif self.mode == "teams":
            return await self._teams_approval(request)
        else:
            return await self._console_approval(request)

    def _auto_approve(self, request: ApprovalRequest) -> ApprovalResult:
        """Automatically approves (for testing/CI)."""
        logger.info(
            "Auto-approving checkpoint %d (%s)",
            request.checkpoint_number,
            request.phase_name,
        )
        return ApprovalResult(
            decision="approved",
            decided_by="AutoApprove (Test Mode)",
            comments="Automatically approved in test mode",
            decided_at=datetime.now(timezone.utc).isoformat(),
            artifact_hash=request.artifact_hash,
        )

    async def _console_approval(self, request: ApprovalRequest) -> ApprovalResult:
        """Interactive console approval for development."""

        separator = "=" * 60
        print()
        print(separator)
        print(f"  CHECKPOINT {request.checkpoint_number}: {request.phase_name}")
        print(separator)
        print()
        print("Artifact Summary (first 1000 chars):")
        print("-" * 40)
        print(request.artifact_summary[:1000])
        print("-" * 40)
        print()
        print(f"Artifact Hash (SHA-256): {request.artifact_hash[:16]}...")
        print(f"Run ID: {request.run_id}")
        print()
        print("Decision:")
        print("  [A] Approve — continue to next phase")
        print("  [R] Reject  — stop pipeline for review")
        print("  [M] Modify  — request changes and re-run phase")
        print()

        while True:
            choice = input("> ").strip().upper()
            if choice in ("A", "APPROVE"):
                decision = "approved"
                break
            elif choice in ("R", "REJECT"):
                decision = "rejected"
                break
            elif choice in ("M", "MODIFY"):
                decision = "modified"
                break
            else:
                print("Invalid choice. Enter A, R, or M.")

        comments = input("Comments (optional, press Enter to skip): ").strip()

        print()
        logger.info(
            "Checkpoint %d: %s (by Console User)",
            request.checkpoint_number,
            decision,
        )

        return ApprovalResult(
            decision=decision,
            decided_by="Console User",
            comments=comments,
            decided_at=datetime.now(timezone.utc).isoformat(),
            artifact_hash=request.artifact_hash,
        )

    async def _teams_approval(self, request: ApprovalRequest) -> ApprovalResult:
        """
        Sends approval request to Microsoft Teams via webhook
        and polls for a response.
        """
        if not self.teams_webhook_url:
            logger.warning("Teams webhook URL not configured. Falling back to console.")
            return await self._console_approval(request)

        # Build Adaptive Card payload
        card_payload = {
            "type": "message",
            "attachments": [
                {
                    "contentType": "application/vnd.microsoft.card.adaptive",
                    "content": {
                        "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
                        "type": "AdaptiveCard",
                        "version": "1.5",
                        "body": [
                            {
                                "type": "TextBlock",
                                "text": f"ForgeSquad Checkpoint {request.checkpoint_number}",
                                "weight": "Bolder",
                                "size": "Large",
                            },
                            {
                                "type": "TextBlock",
                                "text": f"Phase: {request.phase_name}",
                                "weight": "Bolder",
                            },
                            {
                                "type": "TextBlock",
                                "text": f"Run ID: {request.run_id}",
                                "isSubtle": True,
                            },
                            {
                                "type": "TextBlock",
                                "text": "Artifact Summary:",
                                "weight": "Bolder",
                                "separator": True,
                            },
                            {
                                "type": "TextBlock",
                                "text": request.artifact_summary[:500],
                                "wrap": True,
                            },
                            {
                                "type": "FactSet",
                                "facts": [
                                    {
                                        "title": "Hash",
                                        "value": request.artifact_hash[:16] + "...",
                                    },
                                    {
                                        "title": "Timeout",
                                        "value": f"{request.timeout_minutes} min",
                                    },
                                ],
                            },
                        ],
                        "actions": [
                            {
                                "type": "Action.Submit",
                                "title": "Approve",
                                "style": "positive",
                                "data": {
                                    "decision": "approved",
                                    "checkpoint": request.checkpoint_number,
                                    "run_id": request.run_id,
                                },
                            },
                            {
                                "type": "Action.Submit",
                                "title": "Reject",
                                "style": "destructive",
                                "data": {
                                    "decision": "rejected",
                                    "checkpoint": request.checkpoint_number,
                                    "run_id": request.run_id,
                                },
                            },
                        ],
                    },
                }
            ],
        }

        # Send to Teams
        async with httpx.AsyncClient() as client:
            try:
                response = await client.post(
                    self.teams_webhook_url,
                    json=card_payload,
                    timeout=30.0,
                )
                response.raise_for_status()
                logger.info(
                    "Teams notification sent for checkpoint %d",
                    request.checkpoint_number,
                )
            except httpx.HTTPError as e:
                logger.error("Failed to send Teams notification: %s", e)
                logger.info("Falling back to console approval")
                return await self._console_approval(request)

        # Poll for response via approval API
        if not self.approval_api_url:
            logger.warning(
                "Approval API URL not configured. "
                "Falling back to console for response."
            )
            return await self._console_approval(request)

        import asyncio

        poll_interval = 30  # seconds
        max_polls = (request.timeout_minutes * 60) // poll_interval

        async with httpx.AsyncClient() as client:
            for _ in range(max_polls):
                try:
                    resp = await client.get(
                        f"{self.approval_api_url}/approvals/{request.run_id}/{request.checkpoint_number}",
                        timeout=10.0,
                    )
                    if resp.status_code == 200:
                        data = resp.json()
                        if data.get("is_completed"):
                            return ApprovalResult(
                                decision=data.get("decision", "rejected"),
                                decided_by=data.get("decided_by", "Teams User"),
                                comments=data.get("comments", ""),
                                decided_at=datetime.now(timezone.utc).isoformat(),
                                artifact_hash=request.artifact_hash,
                            )
                except httpx.HTTPError:
                    pass  # Continue polling

                await asyncio.sleep(poll_interval)

        # Timeout
        logger.warning(
            "Checkpoint %d timed out after %d minutes",
            request.checkpoint_number,
            request.timeout_minutes,
        )
        return ApprovalResult(
            decision="rejected",
            decided_by="System (Timeout)",
            comments=f"Approval timed out after {request.timeout_minutes} minutes",
            decided_at=datetime.now(timezone.utc).isoformat(),
            artifact_hash=request.artifact_hash,
        )


def verify_audit_trail(audit_trail: list[dict]) -> dict:
    """
    Verifies the integrity of an audit trail by checking the hash chain.

    Args:
        audit_trail: List of audit events in chronological order.

    Returns:
        Verification result with valid flag and any issues found.
    """
    if not audit_trail:
        return {"valid": False, "message": "Empty audit trail"}

    issues: list[str] = []
    expected_prev = "genesis"

    for i, event in enumerate(audit_trail):
        # Check previous hash chain
        if event.get("previous_hash") != expected_prev:
            issues.append(
                f"Event {i}: previous hash mismatch "
                f"(expected {expected_prev[:8]}..., got {event.get('previous_hash', 'N/A')[:8]}...)"
            )

        # Recompute event hash
        content = (
            f"{event['run_id']}|{event['event_type']}|{event['phase']}|"
            f"{event['agent']}|{event['message']}|{event.get('previous_hash', 'genesis')}"
        )
        computed = hashlib.sha256(content.encode()).hexdigest()

        if event.get("event_hash") != computed:
            issues.append(
                f"Event {i}: hash mismatch "
                f"(stored {event.get('event_hash', 'N/A')[:8]}..., "
                f"computed {computed[:8]}...)"
            )

        expected_prev = event.get("event_hash", "")

    return {
        "valid": len(issues) == 0,
        "total_events": len(audit_trail),
        "issues": issues,
        "message": (
            "Audit trail integrity verified"
            if not issues
            else f"INTEGRITY VIOLATION: {len(issues)} issues found"
        ),
    }
